src/private/functions.php

```diff
@@ -0,0 +1,121 @@
+<?php
+
+namespace DDTrace\Private_;
+
+use DDTrace\Http\Urls;
+
+// Constants definition with [] content is not allowed in 5.4, so we need a class for 5.4 compatibility.
+class Constants
+{
+    public static function getDefaultUriPathNormalizeRegexes()
+    {
+        return [
+            '/^\d+$/',
+            '/^[0-9a-f]{8}-?[0-9a-f]{4}-?[1-5][0-9a-f]{3}-?[89ab][0-9a-f]{3}-?[0-9a-f]{12}$/',
+            '/^[0-9a-f]{8,128}$/',
+        ];
+    }
+}
+
+/**
+ * Given a uri path in the form '/user/123/path/Name' it returns a normalized path applying the correct outgoing rules:
+ * e.g. '/user/?/path/?'
+ * Note: it also accepts full urls which are preserved: http://example.com/int/123 ---> http://example.com/int/?
+ *
+ * @param string $uriPath
+ * @return string
+ */
+function util_uri_normalize_outgoing_path($uriPath)
+{
+    return _util_uri_apply_rules($uriPath, /* incoming */ false);
+}
+
+/**
+ * Given a uri path in the form '/user/123/path/Name' it returns a normalized path applying the correct incoming rules:
+ * e.g. '/user/?/path/?'
+ *
+ * @param string $uriPath
+ * @return string
+ */
+function util_uri_normalize_incoming_path($uriPath)
+{
+    return _util_uri_apply_rules($uriPath, /* incoming */ true);
+}
+
+/**
+ * @param string $uriPath
+ * @param boolean $incoming
+ * @return string
+ */
+function _util_uri_apply_rules($uriPath, $incoming)
+{
+    if ('/' === $uriPath || '' === $uriPath || null === $uriPath) {
+        return '/';
+    }
+
+    // We always expect leading slash if it is a pure path, otherwise if it is a full url we preserve schema and port.
+    if ($uriPath[0] !== '/' && substr($uriPath, 0, 4) !== "http") {
+        $uriPath = '/' . $uriPath;
+    }
+
+    $fragmentRegexes = \ddtrace_config_path_fragment_regex();
+    $incomingMappings = \ddtrace_config_path_mapping_incoming();
+    $outgoingMappings = \ddtrace_config_path_mapping_outgoing();
+
+    // We can now be in one of 3 cases:
+    //   1) At least one among DD_TRACE_RESOURCE_URI_FRAGMENT_REGEX and DD_TRACE_RESOURCE_URI_MAPPING_INCOMING|OUTGOING
+    //      is defined. Then ignore legacy DD_TRACE_RESOURCE_URI_MAPPING and apply *new normalization*.
+    //   2) Only DD_TRACE_RESOURCE_URI_MAPPING is defined, then apply *legacy normalization* for backward compatibility.
+    //   3) Nothing is defined, then apply *new normalization*.
```

Comment:
Just to confirm the behavior: as of right now if someone is currently using `DD_TRACE_RESOURCE_URI_MAPPING` to configure all their incoming URL's and then set `DD_TRACE_RESOURCE_URI_MAPPING_OUTGOING` after this release, the incoming URL's will no longer get normalized. If that is the case we should definitely make that clear in the documentation. :)

Reply:
This is the behavior that I proposed.

> If that is the case we should definitely make that clear in the documentation

Absolutely, I am currently working on the docs PR. Do you agree with the behavior?

Reply:
I'm good with the behavior given good docs. 😄